EIPS/eip-6963.md

```diff
@@ -0,0 +1,211 @@
+
+A locally unique identifier prevents from conflicts using the same globally unique identifier by using UUID v4.0
+
+A globally unique identifier is used to for machine-readable detection of which wallet is being used and it can take different formats, for example:
+
+```sh
+# lowercase name
+awesomewallet
+
+# legacy JS variable
+isCoolWallet
+
+# reserved domain
+io.dopewallet.app
+```
+
+A uri encoded image was chosen to enable flexibility for multiple protocols for fetching and rendering icons, for example:
+
+```sh
+# svg (data uri)
+data:image/svg+xml,<svg version="1.1" xmlns="http://www.w3.org/2000/svg" width="32px" height="32px" viewBox="0 0 32 32"><circle fill="red" cx="16" cy="16" r="12"/></svg>
+# png (data uri)
+data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAUAAAAFCAYAAACNbyblAAAAHElEQVQI12P4//8/w38GIAXDIBKE0DHxgljNBAAO9TXL0Y4OHwAAAABJRU5ErkJggg==
+# jpg (ipfs uri)
+ipfs://QmZ8ify1Z3pSpJjRANKsWLDqZrJjjZJ9e7o65LheQ2pFqb
+# webp (http uri)
+https://ethereum.org/static/f541df14fca86543040c113725b5bd1a/99bcf/metamask.webp
+```
+
+Regarding the event-based approach, previous proposal introduced mechanisms that relied on a single window object that could be overriden by multiple parties
+
+Therefore using an event-based approach we avoid these race conditions and potential attacks by making the communication available both ways.
+
+## Backwards Compatibility
+
+This EIP doesn't require supplanting `window.ethereum`, so it doesn't directly break existing applications. However, the recommended behavior of eventually supplanting `window.ethereum` would break existing applications that rely on it.
+
+## Reference Implementation
+
+### Wallet Provider
+
+Here is a reference implementation for an injected script by a wallet provider to support this new interface in parallel with the existing pattern.
+
+```typescript
+function onPageLoad() {
+  let ethereum: EIP1193Provider
+
+  window.ethereum = ethereum
+
+  function announceProvider() {
+    const info: ProviderInfo = {
+      walletId: "org.example.wallet"
+      uuid: "350670db-19fa-4704-a166-e52e178b59d2
+      name: "Example Wallet"
+      icon: "https://wallet.example.org/icon.png"
+    }
+    const provider = { info, ethereum }
+    window.dispatchEvent(new CustomEvent("evmProviderAnnounced", { detail: provider }))
+  }
+
+  window.addEventListener("evmProviderRequested", (event: EVMProviderRequestedEvent) => {
+    announceProvider()
+  })
+
+  announceProvider()
+}
+```
+
+### Ethereum library
+
+Here is a reference implementation for a Ethereum library to display and track multiple wallet providers that are injected by browser extensions.
+
+```typescript
+function onPageLoad() {
+  const providers: EVMProvider[] = [];
+
+  window.addEventListener(
+    "evmProviderAnnounced",
+    (event: EVMProviderAnnouncedEvent) => {
+      providers.push(event.detail);
+    }
+  );
+
+  window.dispatchEvent(new Event("evmProviderRequested"));
```

Comment:
Can we provide guidance on what to do if a browser extension injects itself after this has been invoked?